### PR TITLE
Router: allowFolding for trailingCommas != never

### LINKED ISSUE #2755: Control newlines with trailing commas
The Dart code formatter `dartfmt` has a very useful feature, which I'm missing in `scalafmt` now:

If there is a trailing comma, the formatter adds line-breaks after every comma. If there isn't, it tries to put all elements into one line and wrap after `max_columns`.
(https://flutter.dev/docs/development/tools/formatting#using-trailing-commas)

You can try this behavior online here:
https://dartpad.dev/

With the following code:
```dart
// try adding and removing the trailing commas, then press the "Format" button

int multiply(
  int x,
  int y,
) =>
    x * y;

int longFunctionName(int a, int b, int c, int d, int e, int f, int g, int h,
        int i, int j, int k, int l) =>
    a + b;
```

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #2755.

### CODE REVIEW

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
Comment (tgodzik) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:1388]: What about `TrailingCommas.always` shouldn't the behaviour be the same?

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:1388]: then we need to add another label. in reality, what they all mean is, if there's a newline after the last param/arg, what should be done. none of them forced that newline, before this change.

Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:1388]: But currently with `TrailingCommas.always` we will remove the comma for anything shorter and add it for anything longer than a single line, which is unexpected. We should at least not remove the comma in that case.

Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:1388]: I am not even talking about when it's on the same line.

Let's say we have:
```scala
apply(a, b,
)
```

with `keep` we will get:
```
apply(
  a,
  b,
)
```

but with always we will get in rewritten to:

```
apply(a, b)
```

This changes if the line is longer so at the same moment another normal method might be rewritten from:
```scala
veryLongMethodName(a, b)
```

to 

```scala
veryLongMethodName(
  a,
  b,
)
```

This can even happen in the same file, two operations doing the apposite things.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:1388]: when i talked about "same line", i meant _after_ scalafmt chose the splits (not in the source code). if we want to force a trailing comma, we also need to force a newline, and i am going to add an explicit parameter for that.

**Review by tgodzik — APPROVED** [review]
LGTM!

**Comment by kitbellew** [comment]
@tgodzik adding support for forcing a comma for `always` and `multiple` turned out to be a bit complex, with 10 commits. i have split them up into several pull requests (see #2758, #2759, #2760).

will not merge this one now, will update it after the other ones are merged.

**Comment by kitbellew** [comment]
@tgodzik rebased on the PRs merged earlier today, implementing the logic via the `allowFolding` parameter and not just for `keep`.

File: scalafmt-tests/src/test/resources/trailing-commas/trailingCommasPreserve.stat
Comment (tgodzik) [inline scalafmt-tests/src/test/resources/trailing-commas/trailingCommasPreserve.stat:210]:
```suggestion
```

This is probably not neccessary.

**Review by tgodzik — APPROVED** [review]
LGTM, just one minor comment, can be ignored.

File: scalafmt-tests/src/test/resources/trailing-commas/trailingCommasPreserve.stat
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/trailing-commas/trailingCommasPreserve.stat:210]: you're right, there are no imports but added anyway for consistency, in case the test is amended in the future, as the title of the test says "binpack".

**Pull request merged**
